### ISSUE TITLE
Support collections of executables in assertAll()

### ISSUE THREAD

**Issue by JLLeitschuh** [junit-team/junit5#1163]
## Overview

Continuing the discussion that was spawned from this Twitter chat:
https://twitter.com/relizarov/status/930828269611401216

To start this discussion, these are the methods exposed by [AssertAll](https://github.com/junit-team/junit5/blob/master/junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertAll.java).

The `assertAll` method does not have an overload that takes a `Collection<Executable>`.

Currently to use `assertAll` from Kotlin you must first turn it into a Java8 `Stream<() -> Unit>`

This is the current example API:
```kotlin
@Test
fun `assertions from a stream`() {
    assertAll(
        "people with name starting with J",
        people
            .stream()
            .map {
                // This mapping returns Stream<() -> Unit>
                { assertTrue(it.firstName.startsWith("J")) }
            }
    )
}
```

I think what @elizarov was proposing in his tweet was to change the API so you could do something like this:

```kotlin
@Test
fun `assertions from a stream`() {
    assertAll(
        "people with name starting with J",
        people
            .map {
                // This mapping returns Stream<() -> Unit>
                { assertTrue(it.firstName.startsWith("J")) }
            }
    )
}
```

Currently the API exposes this method in kotlin:
```kotlin
/**
 * [Stream] of functions to be executed.
 */
private typealias ExecutableStream = Stream<() -> Unit>
private fun ExecutableStream.convert() = map { Executable(it) }

/**
 * @see Assertions.assertAll
 */
fun assertAll(executables: ExecutableStream) =
    Assertions.assertAll(executables.convert())

/**
 * @see Assertions.assertAll
 */
fun assertAll(heading: String?, executables: ExecutableStream) =
    Assertions.assertAll(heading, executables.convert())
```

The proposed suggestion would add additional assertion methods:
```kotlin
/**
 * [Collection] of functions to be executed.
 */
private typealias ExecutableCollection = Collection<() -> Unit>
/**
 * @see Assertions.assertAll
 */
fun assertAll(executables: ExecutableCollection) =
    assertAll(executables.stream())
/**
 * @see Assertions.assertAll
 */
fun assertAll(heading: String?, executables: ExecutableCollection) =
    assertAll(heading, executables.stream())
```

If the team decides that these methods make sense to be added, should be reflected in the Java API?
The additions would be something like this:

```java
static void assertAll(Collection<Executable> executables) {
    assertAll(null, executables);
}

static void assertAll(String heading, Collection<Executable> executables) {
    assertAll(heading, executables.stream());
}
```

## Question for the team

Why was `void assertAll(Collection<Executable> executables)` not added to the API to begin with?
Do those same reasons apply here? Do they only apply to Java but not Kotlin?

## Deliverables

- [x] Support `Collection<Executable>` in Java for `assertAll()`.
- [x] Document `Collection<Executable>` support in Java in Release Notes.
- [ ] Support `Collection<() -> Unit>` in Kotlin for `assertAll()`.
- [ ] Document `Collection<() -> Unit>` support in Kotlin in Release Notes (if necessary).
- [ ] Update Kotlin examples in User Guide.

**Comment by sbrannen** [comment]
> Why was `void assertAll(Collection<Executable> executables)` not added to the API to begin with?

We originally only supported `Executable...`.

The reason we support `Stream<Executable>` is simply because someone asked for that support.

So, the answer to your question is that we didn't foresee a need for supporting `Collection<Executable>`.

However, I don't see any problem with adding `assertAll(Collection<Executable>)` to the API. We'll simply convert it internally and delegate to existing methods.

**Comment by sbrannen** [comment]
_in progress_

**Comment by sbrannen** [comment]
@JLLeitschuh I'll push the changes for Java to `master`, and I've assigned this issue to you as well in order for you to make the necessary changes to the Kotlin API.

Deal?

**Comment by sbrannen** [comment]
On second thought, if we go this route, wouldn't it be better to just support `Iterable<Executable>` instead of `Collection<Executable>`?

@JLLeitschuh would `Iterable<Executable>` work just as well in the Kotlin API?

**Comment by sbrannen** [comment]
Never mind... `Iterable<Executable>` doesn't even work in Java since it _clashes_ with `Executable...` which is an array which is also an `Iterable`. 

So we'll just go with `Collection`.

**Comment by JLLeitschuh** [comment]
😄 
I'd be happy to implement the kotlin side unless you want to take a crack at writing some kotlin code yourself. Either way is fine.

**Comment by sbrannen** [comment]
@JLLeitschuh We'd like to get 5.1 M1 out the door as soon as possible (perhaps as early as tonight).

So, with that in mind, if you could implement the necessary changes for Kotlin ASAP, that'd be awesome!

FYI: I think I'll be pushing the necessary changes on the Java side within the next few minutes. :wink:

**Comment by JLLeitschuh** [comment]
I'll try to get to it before Noon EST. I have ~ an hour drive to work and I'm about to leave.

I'll try to do it ASAP and hopefully won't hit any hiccups along the way.

**Comment by jbduncan** [comment]
@sbrannen 

> Never mind... `Iterable<Executable>` doesn't even work in Java since it _clashes_ with `Executable...` which is an array which is also an `Iterable`.

Oh, really? I knew that arrays could be subjects for foreach loops, but I believe that this code example is actually illegal?

```java
public static void main(String[] args) {
  iterateAndPrint(args);
}

private static void iterateAndPrint(Iterable<String> values) {
  values.forEach(System.out::println);
}
```

I'm sure I've misunderstood something here. Would you mind clarifying things further for me?

**Comment by sormuras** [comment]
Java Shell example works:
```
jshell
|  Welcome to JShell -- Version 9.0.1
|  For an introduction type: /help intro

jshell> void all(Runnable... args) { for(Runnable a : args) a.run(); }
|  created method all(Runnable...)

jshell> void all(Iterable<Runnable> iter) { for(Runnable a : iter) a.run(); }
|  created method all(Iterable<Runnable>)

jshell> all(() -> System.out.println("Hi"))
Hi

jshell> all(List.of(() -> System.out.println("Hi")))
Hi
```

 Sam tried it on master with the `assertAll` and one more tests failed:
```
/Users/sbrannen/source/junit5/junit-jupiter-engine/src/test/java/org/junit/jupiter/api/AssertAllAssertionsTests.java:133: error: unreported exception IOException; must be caught or declared to be thrown
                        throw new IOException();
```

**Comment by sbrannen** [comment]
**Update:**

The Java side of things has been addressed in 61ce1f71befaf00fe3bf0e709a6d78fd3986f726 in `master`.

**Comment by sbrannen** [comment]
> I'll try to do it ASAP and hopefully won't hit any hiccups along the way.

@JLLeitschuh Thanks!

**Comment by sbrannen** [comment]
> jshell> all(() -> System.out.println("Hi"))

@sormuras Did you try that with a lambda expression that throws an exception???

**Comment by sbrannen** [comment]
The exceptions are what are causing problems for us somehow.

**Comment by sbrannen** [comment]
> Sam tried it on master with the assertAll and one more tests failed:

Minor correction: the tests actually do not _compile_. :wink:

**Comment by JLLeitschuh** [comment]
With this change you have to do this which is still kinda weird (the doubly nested braces `{ { } }`.
Eg:
```kotlin
@Test
fun `assertions from a collection`() {
    assertAll(
        "people with last name of Doe",
        people.map { { assertEquals("Doe", it.lastName) } }
    )
}
```
One thing I was just thinking about that might make the API even nicer for kotlin would be exposing extension methods that would let you do something like this:

```kotlin
@Test
fun `assertions from a collection`() {
    assertAll(
        "people with last name of Doe",
        people.mapExecutable { assertEquals("Doe", it.lastName) }
    )
}
```

The signature of that would be something like this:
```kotlin
package org.junit.jupiter.api

public inline fun <T, () -> Unit> Iterable<T>.mapExecutable(transform: (T) -> Unit): Collection<() -> Unit> {
    return map { { transform(it) } }
}
```
This might go beyond the scope of the API that the team wants to expose.
@sbrannen or @elizarov what are your thoughts?

**Comment by sormuras** [comment]
Last `jshell` example -- with `Callable` now:
```
jshell> /list

   1 : void all(Callable... args) { for (Callable a : args) { try { System.out.println(a.call()); } catch
       (Throwable t) { } }
       }
   2 : all(() -> 1)
   3 : all(() -> "10")
   5 : void all(Iterable<Callable> iter) { for (Callable a : iter) { try { System.out.println(a.call()); } catch
       (Throwable t) { } } }
   6 : all(() -> 2)
   7 : all(Set.of(() -> 2))

jshell> all(() -> 2)
2

jshell> all(Set.of(() -> 2))
2
```

Will try `assertAll(Iterable<Executable> iterable)` soon.

**Comment by sbrannen** [comment]
> This might go beyond the scope of the API that the team wants to expose.
@sbrannen or @elizarov what are your thoughts?

@JLLeitschuh, can you please open a new issue to discuss that post 5.1 M1?

**Comment by sbrannen** [comment]
> Oh, really? I knew that arrays could be subjects for foreach loops, but I believe that this code example is actually illegal?

@jbduncan, That's correct: your example won't compile.

But I wasn't talking about that use case. Rather, we have overloaded variants of `assertAll()` that accept lambda expressions of type `Executable`, and when I added the new `assertAll(Iterable<Executable>)` variant, code that previously compiled with a single `Executable` lambda expression that threw an exception ceased to compile.

Perhaps it's a bug in the JDK 9 compiler. I don't know: I didn't investigate further.
